src/note.rs

```diff
@@ -0,0 +1,394 @@
+    pub fn transparent(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Transparent, psk, value)
+    }
+
+    /// Creates a new obfuscated note
+    pub fn obfuscated(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Obfuscated, psk, value)
+    }
+
+    /// Create a new phoenix output note without inner randomness
+    pub fn deterministic(
+        note_type: NoteType,
+        r: &JubJubScalar,
+        nonce: JubJubScalar,
+        psk: &PublicSpendKey,
+        value: u64,
+        blinding_factor: JubJubScalar,
+    ) -> Self {
+        let stealth_address = psk.gen_stealth_address(r);
+
+        let value_commitment = JubJubScalar::from(value);
+        let value_commitment = JubJubExtended::from(JubJubAffine::from(
+            &(GENERATOR_EXTENDED * value_commitment)
+                + &(GENERATOR_NUMS_EXTENDED * blinding_factor),
+        ));
+
+        // Output notes have undefined position
+        let pos = 0;
+
+        let encrypted_data = match note_type {
+            NoteType::Transparent => {
+                chunk_of!(ENCRYPTED_DATA_SIZE; &value.to_le_bytes())
+            }
+            NoteType::Obfuscated => {
+                let shared_secret = dhke(r, psk.A());
+                let nonce = BlsScalar::from(nonce);
+                let value = BlsScalar::from(value);
+                let blinding_factor = BlsScalar::from(blinding_factor);
+
+                let mut cipher = PoseidonCipher::encrypt(
+                    &[value, blinding_factor],
+                    &shared_secret,
+                    &nonce,
+                );
+
+                // TODO: replace the internals of PoseidonCipher with a plain
+                // array.
+                // Adding an infallible `to_bytes` method too.
+                let mut encrypted_data = [0u8; 96];
+                cipher
+                    .read_exact(&mut encrypted_data)
+                    .expect("Cannot encode PoseidonCipher");
+                encrypted_data
+            }
+        };
+
+        Note {
+            note_type,
+            value_commitment,
+            nonce,
+            stealth_address,
+            pos,
+            encrypted_data,
+        }
+    }
+
+    fn decrypt_data(&self, vk: &ViewKey) -> (u64, JubJubScalar) {
+        let R = self.stealth_address.R();
+        let shared_secret = dhke(vk.a(), R);
+        let nonce = BlsScalar::from(self.nonce);
+
+        // TODO: replace the internals of PoseidonCipher with a plain
+        // array.
+        let mut cipher = PoseidonCipher::default();
+
+        // TODO: improve the error handling
+        cipher
+            .write(&self.encrypted_data[..])
+            .expect("Cannot write the PoseidonCipher");
+
+        let data = cipher.decrypt(&shared_secret, &nonce);
+
+        // TODO: need to figure out what we want to do in case the decryption
+        // fails
+        return if data.is_err() {
+            (0u64, JubJubScalar::zero())
```

Comment:
We should simply return the error IMO. If the decryption fails you don't have anything to do with the Note.